src/json.hpp

```diff
@@ -144,6 +183,50 @@ struct DecimalSeparator : std::numpunct<char>
 
 }
 
+// taken from ranges-v3
+// TODO add doc
+template <typename T>
+struct _static_const
+{
+  static constexpr T value{};
+};
+
+template <typename T>
+constexpr T _static_const<T>::value;
+
+inline namespace
+{
+  constexpr auto const& to_json = _static_const<detail::to_json_fn>::value;
+  constexpr auto const& from_json = _static_const<detail::from_json_fn>::value;
+}
+
+// default JSONSerializer template argument, doesn't care about template argument
+// will use ADL for serialization
+template <typename = void, typename = void>
+struct adl_serializer
+{
+  template <typename T, typename Json, typename = enable_if_t<std::is_default_constructible<uncvref_t<T>>::value>>
+  static auto from_json(Json&& j) -> uncvref_t<decltype(::nlohmann::from_json(std::forward<Json>(j), std::declval<T&>()), std::declval<T>())>
+  {
+    uncvref_t<T> ret;
+    ::nlohmann::from_json(std::forward<Json>(j), ret);
+    return ret;
+  }
+
+  template <typename T, typename Json>
+  static auto from_json(Json&& j, T& val) -> decltype(::nlohmann::from_json(std::forward<Json>(j), val))
+  {
+    ::nlohmann::from_json(std::forward<Json>(j), val);
+  }
```

Comment:
IIUC  `from_json` should never return a value, that is, its signature should be `(JSON const& input, T& output) -> void`, so maybe it would be better to omit the return value but put a `void` in the `decltype` like this:

```c++
static auto from_json(Json&& j, T& val) -> 
decltype(::nlohmann::from_json(std::forward<Json>(j), val), void)
{
    ::nlohmann::from_json(std::forward<Json>(j), val);
}
```